google-oauth-client/src/main/java/com/google/api/client/auth/oauth2/JWTAuthentication.java

```diff
@@ -0,0 +1,127 @@
+ * <a href="https://tools.ietf.org/html/rfc7523">JSON Web Token (JWT) Profile
+ *       for OAuth 2.0 Client Authentication and Authorization Grants</a>
+ *
+ * <p>This implementation assumes that the {@link HttpRequest#getContent()} is {@code null} or an
+ * instance of {@link UrlEncodedContent}. This is used as the client authentication in {@link
+ * TokenRequest#setClientAuthentication(HttpExecuteInterceptor)}.
+ *
+ * <p>
+ *     To use JWT authentication, grant_type must be "client_credentials".
+ *     If AuthorizationCodeTokenRequest.setGrantType() is called, set it to
+ *     JWTAuthentication.GRANT_TYPE_CLIENT_CREDENTIALS. It can also be left
+ *     uncalled. Setting it to any other value causes an IllegalArgumentException.
+ * </p>
+ *
+ * <p>Sample usage:
+ *
+ * <pre>
+ * static void requestAccessToken() throws IOException {
+ * try {
+ * TokenResponse response = new AuthorizationCodeTokenRequest(new NetHttpTransport(),
+ * new GsonFactory(), new GenericUrl("https://server.example.com/token")
+ * .setGrantType(JWTAuthentication.GRANT_TYPE_CLIENT_CREDENTIALS)
+ * .setClientAuthentication(
+ * new JWTAuthentication("eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzM4NCJ9...")).execute();
+ * System.out.println("Access token: " + response.getAccessToken());
+ * } catch (TokenResponseException e) {
+ * if (e.getDetails() != null) {
+ * System.err.println("Error: " + e.getDetails().getError());
+ * if (e.getDetails().getErrorDescription() != null) {
+ * System.err.println(e.getDetails().getErrorDescription());
+ * }
+ * if (e.getDetails().getErrorUri() != null) {
+ * System.err.println(e.getDetails().getErrorUri());
+ * }
+ * } else {
+ * System.err.println(e.getMessage());
+ * }
+ * }
+ * }
+ * </pre>
+ *
+ * <p>Implementation is immutable and thread-safe.
+ *
+ * @author Jun Ying
+ */
+
+public class JWTAuthentication
+    implements HttpRequestInitializer, HttpExecuteInterceptor {
+
+    public static final String GRANT_TYPE_KEY = "grant_type";
+
+    /** Predefined value for grant_type when using JWT **/
+    public static final String GRANT_TYPE_CLIENT_CREDENTIALS = "client_credentials";
+
+    public static final String CLIENT_ASSERTION_TYPE_KEY = "client_assertion_type";
+
+    /** Predefined value for client_assertion_type when using JWT **/
+    public static final String CLIENT_ASSERTION_TYPE = "urn:ietf:params:oauth:client-assertion-type:jwt-bearer";
+
+    public static final String CLIENT_ASSERTION_KEY = "client_assertion";
+
+    /** JWT for authentication. */
+    private final String jwt;
+
+    /**
+     * @param jwt JWT used for authentication
+     */
+    public JWTAuthentication(String jwt) {
+        this.jwt = Preconditions.checkNotNull(jwt);
+    }
+
+    public void initialize(HttpRequest request) throws IOException {
+        request.setInterceptor(this);
+    }
+
+    public void intercept(HttpRequest request) {
+        Map<String, Object> data = Data.mapOf(UrlEncodedContent.getContent(request).getData());
```

Comment:
This method seems to be mutating the internal state of some object. This is very suspicious.

Reply:
The purpose of an authentication intercept pretty much requires it. It needs to add the appropriate authentication information into the HTTP request. It's the decorator pattern. In any case, it is implemented exactly like ClientParametersAuthentication already in the package.